Altar refinements

wchen1990 opened the pull request:
- Change altar recipe check to check if the inventory and recipe have differing input sizes rather than checking if the inputs are empty to allow for catalysts to be used on the first use rather than needing a dummy item before the catalyst
- Add an indestructible altar for use cases where one would want a permanent fixture that people would visit
- Use DEFAULT_ZONE if region is not defined in the json
- Adjust altar_active.png.mcmeta so that the animation loops more seamlessly

wchen1990 committed:
Altar refinements

- Change altar recipe check to check if the inventory and recipe have differing input sizes rather than checking if the inputs are empty to allow for catalysts to be used on the first use rather than needing a dummy item before the catalyst
- Add an indestructible altar for use cases where one would want a permanent fixture that people would visit
- Use DEFAULT_ZONE if region is not defined in the json
- Adjust altar_active.png.mcmeta so that the animation loops more seamlessly

rlnt commented on src/main/java/com/almostreliable/summoningrituals/Constants.java:
```diff
@@ -26,6 +26,7 @@ public final class Constants {
     public static final String NETWORK = "network";
     public static final String OFFSET = "offset";
     public static final String OUTPUTS = "outputs";
+    public static final String INDESTRUCTIBLE_ALTAR = "indestructible_altar";
```
The file is sorted alphabetically, please sort it.

rlnt commented on src/main/java/com/almostreliable/summoningrituals/altar/AltarBlock.java:
```diff
@@ -38,14 +38,19 @@
-    private static final VoxelShape SHAPE = Stream.of(
+    public static final BooleanProperty ACTIVE = BooleanProperty.create(Constants.ACTIVE);
+    public static final DirectionProperty FACING = BlockStateProperties.HORIZONTAL_FACING;
+    public static final VoxelShape SHAPE = Stream.of(
```
What is the use case of making those `public`? You are not using them anywhere.

wchen1990 replied:
I figured in case someone would want to extend the AltarBlock class in the future it'd make a lot of things easier for them to implement whatever changes they'd need that depends on these states and the methods that I've made protected. But, I suppose that can be done when that use case arises.

rlnt commented on src/main/java/com/almostreliable/summoningrituals/altar/AltarBlock.java:
```diff
@@ -38,14 +38,19 @@
         Block.box(2, 9, 2, 14, 13, 14)
     ).reduce((v1, v2) -> Shapes.join(v1, v2, BooleanOp.OR)).get();
 
+    public AltarBlock(Properties properties) {
+        super(properties);
+        registerDefaultState(defaultBlockState().setValue(FACING, Direction.NORTH).setValue(ACTIVE, false));
+    }
```
If you need a constructor like that, you can get rid of the other one and pass the block properties from the `Setup` class instead.

rlnt commented on src/main/java/com/almostreliable/summoningrituals/altar/AltarBlock.java:
```diff
@@ -135,7 +140,7 @@ protected void createBlockStateDefinition(Builder<Block, BlockState> builder) {
     }
 
-    private void renderCandleActive(Level level, int x, int y, int z, Vector3f vec) {
+    protected void renderCandleActive(Level level, int x, int y, int z, Vector3f vec) {
```
Not necessary

rlnt commented on src/main/java/com/almostreliable/summoningrituals/altar/AltarBlock.java:
```diff
@@ -147,7 +152,7 @@ private void renderCandleActive(Level level, int x, int y, int z, Vector3f vec)
     }
 
-    private void renderCandleInactive(Level level, int x, int y, int z, Vector3f vec) {
+    protected void renderCandleInactive(Level level, int x, int y, int z, Vector3f vec) {
```
Not necessary

rlnt commented on src/main/java/com/almostreliable/summoningrituals/altar/IndestructibleAltarBlock.java:
```diff
@@ -0,0 +1,10 @@
+package com.almostreliable.summoningrituals.altar;
+
+import net.minecraft.world.level.block.SoundType;
+import net.minecraft.world.level.material.Material;
+
+public class IndestructibleAltarBlock extends AltarBlock{
+    public IndestructibleAltarBlock() {
+        super(Properties.of(Material.STONE).strength(-1.0F, 3600000.0F).sound(SoundType.STONE));
+    }
+}
```
This should not be another class. You can make use of the `AltarBlock` class with your new constructor. Remove this class and pass the block properties directly from `Setup`.

rlnt commented on src/main/java/com/almostreliable/summoningrituals/recipe/AltarRecipe.java:
```diff
@@ -56,7 +56,7 @@ public class AltarRecipe implements Recipe<AltarInvWrapper> {
     @Override
     public boolean matches(AltarInvWrapper inv, Level level) {
-        if (inv.getCatalyst().isEmpty() || inv.getInputs().isEmpty() || !catalyst.test(inv.getCatalyst())) {
+        if (inv.getCatalyst().isEmpty() || inv.getInputs().size() != this.inputs.size() || !catalyst.test(inv.getCatalyst())) {
```
This change will break the crafting logic if you have more items in the Altar than you need since the Altar would require the exact amount of items the recipe has.

wchen1990 replied:
In that case, I'll just remove the input check, since having items in the altar when a recipe only requires the catalyst plus sacrifices feels awkward. Let me know if that's an issue.

rlnt commented on src/main/java/com/almostreliable/summoningrituals/recipe/component/RecipeSacrifices.java:
```diff
@@ -40,7 +40,9 @@ public static RecipeSacrifices fromJson(JsonObject json) {
             sacrifices.add(Sacrifice.fromJson(entity.getAsJsonObject()));
         }
-        var zone = SerializeUtils.vec3FromJson(json.getAsJsonObject(Constants.REGION));
+        var zone = (json.has(Constants.REGION)) ?
```
Remove the brackets around the ternary condition.

wchen1990 committed:
Conform to established coding style

- Undo access changes to AltarBlock
- Use the new AltarBlock constructor for instantiating
- Moved INDESTRUCTIBLE_ALTAR to its appropriate place in Constants class
- Removed input size check for use cases where the ritual has excess items stored for repeated usages of a ritual

DAmNRelentless commented:
I am going to assume you removed the input requirement to allow recipes without any items at all? So if you only want sacrifices for example?

wchen1990 commented:
Yup, that's it.

Mentioned in the comments under your review, but yea when I was testing the mod and made a sacrifice only recipe it required a filler item before the catalyst could be used which felt a bit awkward.

rlnt committed:
format

rlnt committed:
fix right-clicking behavior on indestructible altar

rlnt committed:
render indestructible altar as cutout

rlnt committed:
update changelog

rlnt reviewed (approved):
lgtm, thanks for your contribution
